### ISSUE TITLE
Strange formatting for closure with match

### ISSUE THREAD

**Issue by jonhoo** [rust-lang-nursery/rustfmt#1395]
Take the following (manually formatted) code:
```rust
fn main() {
    let bar = Some(true);
    let foo = Some(true);
    let mut x = false;
    bar.and_then(|_| {
        match foo {
            None => None,
            Some(b) => {
                x = true;
                Some(b)
            }
        }
    });
}
```
`rustfmt` will currently turn that into
```rust
fn main() {
    let bar = Some(true);
    let foo = Some(true);
    let mut x = false;
    bar.and_then(|_| match foo {
                     None => None,
                     Some(b) => {
        x = true;
        Some(b)
    }
                 });
}
```
Which is much less visually clear. It almost seems like `rustfmt` forgets to carry along the indentation level?

**Comment by jonhoo** [comment]
I wonder if this is just another manifestation of #1332...

**Comment by aldanor** [comment]
#1411 could be the the same thing?

**Comment by jonhoo** [comment]
@aldanor in fact, I think these are all the result of the new visual indent rules for closures (in #1332). If you look at these cases, rustfmt does "the right thing" for all of them, in the sense of choosing to use visual indent for the closure body, and then block closure for some things *inside* the closure body. The bug seems to be that block indent inside a visual indent does not inherit the indentation level of the containing visual indent.

**Comment by nrc** [comment]
With the RFC style, this is currently formatted as:

```
fn main() {
    let bar = Some(true);
    let foo = Some(true);
    let mut x = false;
    bar.and_then(
        |_| match foo {
            None => None,
            Some(b) => {
                x = true;
                Some(b)
            }
        },
    );
}
```

Which is much better, but not quite perfect.

**Comment by aldanor** [comment]
Better indeed. However `match` lining up with the match arms (`None` and `Some`) is particularly unfortunate, since `|_| ` happens to take exactly 4 spaces...

In the original example, closure body was surrounded by braces -- were they removed by `rustfmt`?

**Comment by nrc** [comment]
> In the original example, closure body was surrounded by braces -- were they removed by rustfmt?

Yes

**Comment by topecongiro** [comment]
Closed by #1454.
